torch/cuda/amp/amp_scaler.py

```diff
@@ -0,0 +1,292 @@
+import torch
+from collections import defaultdict
+from torch._six import container_abcs
+
+
+class AmpScaler(object):
+    """
+    Performs dynamic gradient scaling, intended to be used when gradients
+    have a dtype with limited dynamic range (e.g. if part of your network is ``torch.float16``).
+
+    Here's how that looks in a simple example::
+
+        scaler = AmpScaler()
+
+        for input, target in data:
+            optimizer.zero_grad()
+            output = model(input)
+            loss = loss_fn(output, target)
+            scaler.scale(loss).backward()
+            scaler.step(optimizer)
+            scaler.update()
+
+    See :ref:`Gradient Scaling Examples<gradient-scaling-examples>` for usage in more complex cases.
+
+    By default, ``scaler.step`` internally unscales ``optimizer``'s gradients before applying them, so the
+    learning rate and other hyperparameters don't need to change.  If you wish to unscale gradients manually
+    prior to :meth:`step`, use :meth:`unscale`.
+
+    ``scaler`` maintains the scale factor internally.  To leverage the dtype's full dynamic range,
+    ``scaler`` attempts to use the largest scale factor it can without incurring overflow.
+    It does so by checking the gradients for infs and NaNs during every :meth:`step` or separate :meth:`unscale`.
+    If no infs/NaNs are found, :meth:`step` runs the underlying ``optimizer.step()`` as usual and
+    :meth:`update` increases the scale factor slightly.  If infs/NaNs are found, :meth:`step` skips the underlying
+    ``optimizer.step()`` (so the params themselves remain unpolluted) and decreases the scale factor.
+
+    Often, ``scaler.step`` skips the underlying ``optimizer.step()`` for the first few iterations
+    as the scale value calibrates.  After that, step skipping should occur rarely (once every few hundred iterations).
+
+    Arguments:
+        init_scale (float, optional, default=2.**24):  Initial scale factor.
+        growth_factor (float, optional, default=1.001):  Factor by which the scale is multiplied during
+            :meth:`update` if no inf/NaN gradients were found this iteration.
+        backoff_factor (float, optional, default=0.5):  Factor by which the scale is multiplied during
+            :meth:`update` if inf/NaN gradients were found this iteration.
+        enabled (bool,optional, default=True):  If ``False``, disables gradient scaling. :meth:`step` simply
+            invokes the underlying ``optimizer.step()``, and other methods become no-ops.
+    """
+    def __init__(self,
+                 init_scale=2.**24,
+                 growth_factor=1.001,
+                 backoff_factor=0.5,
+                 enabled=True):
+        self._enabled = enabled
+        if enabled:
+            self._scale = torch.full((1,), init_scale, dtype=torch.float32, device="cuda")
+            self._growth_factor = 1.001
```

Comment:
self._growth_factor = growth_factor. You are ignoring your inputs.

Reply:
:man_facepalming: thanks!